src/core/storage/fileio/s3_filesys.hpp

```diff
@@ -0,0 +1,384 @@
+/* Copyright © 2017 Apple Inc. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-3-clause license that can
+ * be found in the LICENSE.txt file or at
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+#ifndef TURI_S3_FILESYS_HPP_
+#define TURI_S3_FILESYS_HPP_
+
+#ifndef TC_DISABLE_REMOTEFS
+
+#include <aws/core/Aws.h>
+#include <aws/s3/S3Client.h>
+#include <aws/s3/model/CreateMultipartUploadRequest.h>
+
+#include <core/storage/fileio/s3_api.hpp>
+#include <string>
+#include <vector>
+
+namespace turi {
+namespace fileio {
+
+namespace s3 {
+
+/*! \brief type of file */
+enum FileType {
+  /*! \brief the file is file */
+  kFile,
+  /*! \brief the file is directory */
+  kDirectory
+};
+
+/*! \brief use to store file information */
+struct FileInfo {
+  /*! \brief full path to the file */
+  s3url path;
+  /*! \brief the size of the file */
+  size_t size;
+  /*! \brief the type of the file */
+  FileType type;
+  /*! \brief default constructor */
+  FileInfo() : size(0), type(kFile) {}
+};
+
+struct ScopedAwsInitAPI {
+  ScopedAwsInitAPI(const Aws::SDKOptions &options) : options_(options) {
+    Aws::InitAPI(options_);
+  }
+  ~ScopedAwsInitAPI() { Aws::ShutdownAPI(options_); }
+  Aws::SDKOptions options_;
+};
+
+const ScopedAwsInitAPI &turi_global_AWS_SDK_setup(
+    const Aws::SDKOptions &options = Aws::SDKOptions());
+
+class Stream {
+ public:
+  /*!
+   * \brief reads data from a stream
+   * \param ptr pointer to a memory buffer
+   * \param size block size
+   * \return the size of data read
+   */
+  virtual size_t Read(void *ptr, size_t size) = 0;
+  /*!
+   * \brief writes data to a stream
+   * \param ptr pointer to a memory buffer
+   * \param size block size
+   */
+  virtual void Write(const void *ptr, size_t size) = 0;
+
+  /*!
+   * \brief closes the stream
+   */
+  virtual void Close() = 0;
+
+  /*! \brief virtual destructor */
+  virtual ~Stream(void) {}
+  /*!
+   * \brief generic factory function
+   *  create an stream, the stream will close the underlying files upon deletion
+   *
+   * \param uri the uri of the input currently we support
+   *            hdfs://, s3://, and file:// by default file:// will be used
+   * \param flag can be "w", "r", "a"
+   * \param allow_null whether NULL can be returned, or directly report error
+   * \return the created stream, can be NULL when allow_null == true and file do
+   * not exist
+   */
+  static Stream *Create(const char *uri, const char *const flag,
```

Comment:
This should return an std::shared_ptr<Stream> instead of a raw pointer.   Also, any reason the url is const char* instead of const std::string& ?

Reply:
The interface is inherited. Will change.

Reply:
Yeah, the original implementation from dmlcio is C style.